File: frontend/app/src/main/java/com/aiinterview/core/websocket/WebSocketClient.kt
package com.aiinterview.core.websocket

import okhttp3.*
import okio.ByteString
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import javax.inject.Inject
import javax.inject.Singleton

class WebSocketClient @Inject constructor(
    private val okHttpClient: OkHttpClient
) {
    private var webSocket: WebSocket? = null
    private val _events = MutableSharedFlow<WebSocketEvent>(replay = 0, extraBufferCapacity = 64)
    val events: SharedFlow<WebSocketEvent> = _events.asSharedFlow()

    fun connect(url: String, token: String? = null) {
        val finalUrl = if (!token.isNullOrEmpty()) {
            val delimiter = if (url.contains("?")) "&" else "?"
            "$url${delimiter}token=$token"
        } else {
            url
        }
        val request = Request.Builder().url(finalUrl).build()
        webSocket = okHttpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _events.tryEmit(WebSocketEvent.Connected)
            }
            override fun onMessage(webSocket: WebSocket, text: String) {
                _events.tryEmit(WebSocketEvent.Message(text))
            }
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _events.tryEmit(WebSocketEvent.Error(t))
            }
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                _events.tryEmit(WebSocketEvent.Closed(code, reason))
            }
        })
    }

    fun send(json: String): Boolean = webSocket?.send(json) ?: false

    fun send(bytes: ByteString): Boolean = webSocket?.send(bytes) ?: false

    fun disconnect() {
        webSocket?.close(1000, "Normal closure")
        webSocket = null
    }

    fun close(code: Int = 1000, reason: String = "Normal closure") {
        webSocket?.close(code, reason)
        webSocket = null
    }
}

File: web/src/api/websocket.client.ts
export class WebSocketClient {
  private ws: WebSocket | null = null;
  private url: string;
  constructor(url: string) {
    this.url = url;
  }
  connect(onMessage?: (msg: any) => void) {
    this.ws = new WebSocket(this.url);
    this.ws.onopen = () => console.log('WS Connected');
    if (onMessage) {
      this.ws.onmessage = (event) => {
        try {
          onMessage(JSON.parse(event.data));
        } catch {
          onMessage(event.data);
        }
      };
    }
  }
  disconnect() {
    this.ws?.close();
  }
  send(data: any) {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(data));
    }
  }
}